runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,486 @@
+local M = {}
+
+local protocol = require('vim.lsp.protocol')
+local uv = vim.loop
+
+---@private
+---Parses the raw pattern into a number of Lua-native patterns.
+---
+---@param pattern string The raw glob pattern
+---@return table A list of Lua patterns. A match with any of them matches the input glob pattern.
+local function parse(pattern)
```

Comment:
is this some sort of glob implementation? why not use `globpath()` and/or `glob2regpat()` ?

Reply:
Yep, that's exactly what this is. LSP specifies its own syntax for the patterns here: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#pattern.

It looks like `glob()` handles everything except the `[!...]` and `{}` syntax that LSP describes. I can take a look to see if we can leverage that for the rest, but I think some amount of finagling like this is going to be necessary to fill the gaps.